accepted/future-releases/class-modifiers/feature-specification.md

```diff
@@ -850,26 +864,50 @@ An anonymous mixin application class cannot be referenced anywhere except in
+
+We define the following transitive property on declarations:
+
+* A declaration *prohibits inheritance* _(of its implementation)_ if:
+
+  * the declaration is marked `interface` or `final`, or
+  * the declaration extends or mixes in a declaration which prohibits inheritance.
+
+Let then *C* be an anonymous mixin application occurring in a post-feature library, with
+superclass *S* and mixin *M*. 
 
-Let *C* be an anonymous mixin application with superclass *S* and mixin *M*. Then:
+* If either *S* or *M* has a `sealed` modifier, then *C* has an implicit `sealed` modifier.
 
-* If any of *S* or *M* has a `sealed` modifier, *C* is has a `sealed` modifier.
-* Otherwise:
-  * *C* is `abstract`, and
-  * If either of *S* or *M* has a `base` or `final` modifier, then *C* has a `final` modifier.
+* Otherwise *C* is `abstract`, and
 
-_We do not distinguish whether *S* or *M* has `base` or `final` modifiers.
-The modifier on *C* is there to satisfy the requirement that a subtype of a `base` or `final`
-declaration is itself `base`, `final` or `sealed`. The anonymous mixin application class will
-be immediately extended inside the same library, which is allowed by both `base` and `final`.,
-and will not be used for anything else._
+  * If *C* *prevents implementation* and *prohibits inheritance*, 
+    then *C* has an implicit `final` modifier.
 
-Adding `sealed` to an anonymous mixin application class with only one subclass ensures
-that the subclass extending the mixin application class can be used
-in exhaustiveness checking of the sealed superclass.
-This is necessary since the anonymous mixin application class itself cannot be referenced.
+  * Otherwise if *C* *prevents implementation*, then *C* has an implicit `base` modifier.
+
+  * Otherwise if *C* *prohibits inheritance*, then *C* has an implicit `interface` modifier.
+
+  * Otherwise *C* has no modifier.
+
+Adding `sealed` to an anonymous mixin application class, which always has precisely
+one subclass, ensures that the subclass can be used in exhaustiveness checking
+of the sealed superclass.
+
+Adding `final` or `base` satisfies the requirement that a subtype of a `base` or `final`
+declaration is itself `base`, `final` or `sealed`.
+
+Adding `interface` ensures that the mixin application class doesn’t remove a restriction
+on the implementation inherited from *S* or *M*. This makes the “reopen” lint described below
+easier to implement.
+
+Adding these modifiers on the anonymous mixin application class ensures that 
+the anonymous class can mostly be ignored, since it satisfies all possible requirements
+of its superclasses, while still propagating those requirements to its subclass. 
+Treating the anonymous class as having these modifiers allows the algorithms
+used to check that restrictions are satisfied, and for the “reopen” lint described below,
+to treat the anonymouse mixin application class as any other class, without needing
+special cases.
```

Comment:
I like to think about this treatment as the one which will implicitly imply a set of modifiers on an anonymous mixin application and on a `sealed` declaration such that they would not incur a compile-time error and would also not cause an `implicit_reopen` lint, in the case where they had been regular (named, non-sealed) declarations.

In other words, the implicit class modifiers serve to propagate the constraints faithfully. This means that it is easy to check for re-opening declarations by looking at their direct superinterfaces. There is no need to search the entire superinterface graph.

I think the text above says most of this, but it doesn't mention that the set of implicit class modifiers are specifying the _most permissive set of restrictions that will not re-open anything_.

Wouldn't it be helpful to say that?

Reply:
Probably would, yes. "It propagates the restrictions of its super-declarations, and odes not add any further", or similar.



I'd generally like to rephrase everything in terms of semantic properties of classes, rather than adding syntactic modifiers that aren't there in the original program.
(I'm a denotational semanticist at heart.)

Say "prevents implementation". A declaration prevents implementation if:
* It's marked `final` or `base`.
* It's an `enum` declaration.
* It has an immediate superinterface which prevents implementation.

(This is the easy one).
Then we say that it's a compile-time error to implement a declaration which prevents implementation.
And that it's a compile-time error if a `class` or `mixin` declaration prevents implementation,
is declared in a post-feature library, and is not marked `base`, `final` or `sealed`.

But we don't have to actually add `final` or `base`, or even figure out which one an anonymous 
mixin application should get.

Likewise, we can say that a class declaration prevents instantiation/cannot have instances if
* it's marked `abstract`,
* it's marked `sealed`, or
* it's an anonymous mixin application class

Rather than adding `abstract` to `sealed` classes, we lift it to the semantic domain and add the semantic property to both `abstract`-declared classes and sealed classes.

It'll just require rewriting everything. Noooo problem. :)

Reply:
But done.